### DISCUSSION TITLE
How to get mpv to properly play HDR videos from Youtube?

### DISCUSSION THREAD

**Discussion by Skyyblaze** [mpv-player/mpv#14591] [Q&A]
First of all I'm amazed at how great and quick it is to watch Youtube videos with mpv. I have a quake style dropdown terminal always running on my Windows install and it's super convenient just calling it, typing mpv with the link and off I go, thanks to everyone that works on mpv!

I have a RTX 4070 and added mpv to Windows AutoHDR along with this config for general SDR Youtube videos which are nicely "converted" to HDR:

`vo=gpu-next
gpu-context=d3d11
hwdec=dxva2
inverse-tone-mapping=yes
d3d11-output-csp=srgb
target-colorspace-hint=yes`

Now I would like to have a second instance of mpv for videos that are actually in HDR. I thought about creating a second instance and simply naming the executable mpvhdr.exe or similar so that this instance won't be picked up by AutoHDR, also within a different folder so it can have its own config file so I can basically type "mpv link" for SDR videos and "mpvhdr link" for HDR videos.

However reading through the mpv manual I struggle to find the right settings for HDR videos, I've tested it with different videos and Youtube in Microsoft Edge side by side and each time mpv looks wildly off color and brightness-wise to what is displayed in the browser.

Also I'm aware that my SDR --> HDR method might not be the best either but it looks good to me. When I tried to expand SDR to HDR with mpv itself everything looked off as well and like this it comes close to what I see in Edge when I use the RTX HDR enhancement for Chromium browser videos.

Could someone point me into the right direction of the manual perhaps? I already saw some topics about this but following the advices there left me mostly with very dull looking videos that have a kind of gray haze over them.

Thanks for help in advance!

**Comment by geextahslex** [comment]
1. Do you need tonemapping or do you have an HDR Monitor/Display?
2. Show some screenshots, what looks weird/off?

**Reply by Skyyblaze** [reply]
> 1. Do you need tonemapping or do you have an HDR Monitor/Display?
> 
>     2. Show some screenshots, what looks weird/off?

@geextahslex

Thanks for the reply!

1: I do have an HDR display, it's a Coolermaster GP27U Mini LED screen with 1200-1400nits. HDR is enabled in Windows, it works fine in Chromium-based browsers and games.

2: I'll attach a screenshots but I'm not sure how well this goes, I found HDR screenshots troublesome.

This is a test video in Edge:

![grafik](https://github.com/user-attachments/assets/b30c8976-8e82-4450-b86a-e3b0fcdfe80e)

This is my normal mpv instance that's AutoHDR enabled:

![grafik](https://github.com/user-attachments/assets/e690e985-0827-4fdf-87dc-2b3a856d98ce)

This is a minimal mpv instance that isn't allowed for AutoHDR with a minimal config:

![grafik](https://github.com/user-attachments/assets/f1ae8c1d-f74c-4764-902f-72dac6fd8234)

`profile=gpu-hq
vo=gpu-next 
hwdec=auto-copy 
target-colorspace-hint=yes`

As you might be able to see, the palm-tree leaves in Edge have a different, warmer shade of green and the white in the sky also appears warmer in Edge. Though to be honest now looking at these three side-by-side for longer I get the impression that both mpv instances look more "correct" than the one in Edge.

EDIT: Oh wow all screenshots came out super overexposed, would you know how can I properly take HDR screenshots in Windows? I could perhaps try taking pictures with my phone and hope that does the tonemapping better?

**Reply by geextahslex** [reply]
The screenshots are fine, they are just to get an idea of what it looks like.
I would say you shouldn't compare browser youtube playback with mpv or any players because they all have their own "stylised" picture/image. It's like VLC, MPV and MPC-HQ will all look slightly different, different tint etc.
You should take a hdr bluray and compare that with multiple players on your machine and decide what looks best for you. 
Also nvidia, windows and your browser will impact the final picture in some way.
HDR is a convoluted topic on itself. I don't know if you are on Win 10 or 11 but from my understanding 11 is superior with handling HDR content. 

during playback press `I` and `2` and show the screenshot. there should be a list that look like this
![wewer](https://github.com/user-attachments/assets/8c8ffb36-e47e-4abd-b629-c404fefee604)

**Reply by Skyyblaze** [reply]
@geextahslex Alright I see, maybe I'm obsessing over things too much then, I just thought that watching a HDR Youtube video in a Chromium browser would lead to the most "correct" result but I guess like with everything else display related these things can by highly subjective and the picture I get via mpv looks by no means bad.

For reference, here's the screenshot you requested:

![grafik](https://github.com/user-attachments/assets/0221cfb1-3060-4c47-9a64-d7399fea37b6)

**Reply by geextahslex** [reply]
okay there is spline, and you shouldn't use tone mapping if you have a HDR setup, so try this
```
tone-mapping=clip
gamut-mapping-mode=clip
```

**Reply by Skyyblaze** [reply]
@geextahslex Thanks for the continued help! Before I try this, how can I identify "spline"? I wonder if it's caused by AutoHDR ontop of actual HDR, here is the same screenshot with AutoHDR disabled:

![grafik](https://github.com/user-attachments/assets/d7357321-042c-4618-bba8-0528a50229e3)

**Reply by geextahslex** [reply]
No problem. 
What do you mean by "identify"? On your screenshot there is "spline tone map"

**Reply by Skyyblaze** [reply]
Oh that's my instance with AutoHDR enabled though, the second screenshot without AutoHDR doesn't show Spline Tone Map does it?

**Reply by Astralyfuse** [reply]
> would you know how can I properly take HDR screenshots in Windows?

If you still having the same issue screenshoting in hdr, you could use [SKIV](https://github.com/SpecialKO/SKIV)
![image](https://github.com/user-attachments/assets/febd6092-5bac-4fbf-b733-fbab4292448b)
 To try it, you first need to download [SpecialK](https://www.special-k.info/) and don't worry, we don't have to use it on any game. After installing them, go to `C:\Program Files\Special K` and open SKIV.exe. It still is experimental and expect to use keybindings. Great thing about it is, it can screenshot hdr image in any format or if you would like tonemap it to sdr. Windowns Snipping Tool already has a way to screenshot in hdr but SKIV looks miles better. You can minimize the program to system tray and use Windows+Ctrl+Shift+O to screenshot anything on your screen.

**Reply by Skyyblaze** [reply]
Thanks for the tip, that sounds great!

**Comment by Andarwinux** [comment]
1. mpv doesn't support AutoHDR or D3D11VideoProcessor HDR
2. Don't explicitly specify `d3d11-output-csp`
3. Set `target-peak=10000`, and verify no tone-mapping by `tone-mapping-visualizeg=yes`

**Reply by Skyyblaze** [reply]
Thanks for the advice, I followed what you said now and this is how my mpv looks without AutoHDR

![grafik](https://github.com/user-attachments/assets/b2f1551e-b7f3-45f1-9c21-dcf9419a1ac9)

Did I do things correctly?

**Reply by Andarwinux** [reply]
Yes, this is pure HDR passthrough, all tone mapping is processed by DWM and monitor itself, same as Chromium or WMP.

**Reply by Skyyblaze** [reply]
Alright, thanks for the help! Would you know if mpv can do SDR to HDR expansion too? A few commands I found for that ended rather weird.

**Reply by Andarwinux** [reply]
> Alright, thanks for the help! Would you know if mpv can do SDR to HDR expansion too? A few commands I found for that ended rather weird.

```
arget-colorspace-hint=yes
inverse-tone-mapping=yes
target-prim=bt.2020
target-trc=pq
target-peak=1200
```

However, it may be difficult to achieve the same results as Chromium or AutoHDR because the underlying workings are completely different.

**Reply by Skyyblaze** [reply]
Thanks again! I just tried this and while it's the best result I had with mpv so far it still messes up the colors somewhat. I think for now I'll stick with my current setup of two different mpv executables and configs, one for native HDR content and one for SDR content that I pull through AutoHDR with an app called AutoHDRForce that makes Windows enable it for any given .exe file.

The results of AutoHDR + mpv are actually quite remarkable for my (untrained) eye.

**Reply by Andarwinux** [reply]
IIRC Windows AutoHDR does not work for mpv, I guess you are using RTX HDR. For this case, `--target-colorspace-hint=yes --target-peak=auto` is enough, nothing else is needed, for HDR videos, mpv will request HDR swapchain, while for SDR video, mpv will request SDR swapchain, RTX HDR only works with SDR swapchain.

**Reply by Skyyblaze** [reply]
Oh AutoHDR does work with mpv if you use a extra tool: https://github.com/7gxycn08/ForceAutoHDR I can confirm it's working because I can actually adjust the intensity of the effect in the Gamebar Overlay which only unlocks when AutoHDR is actually active and moving the slider has an effect on the mpv window. Since you have to specify an executable for AutoHDR this has the side-effect that I had to setup two mpv instances, I called the second one mpv-hdr.exe for native HDR content and the regular mpv.exe I leave for SDR content and then added that to AutoHDR.

I tried RTX HDR before but it doesn't work properly on my end even in games and has the caveat that it only works in single-monitor mode. For whatever reason whenever I try to enhance something with RTX HDR I either get a washed out image with a gray haze or very low brightness overall. Nvidia stated that they are aware of issues with RTX HDR where the monitor doesn't report the correct peak-brightness in its EDID so hopefully they fix that eventually.

My only success with HDR so far is:

- AutoHDR
- SpecialK HDR Injection
- native HDR in videos/games
